### PR TITLE
allow spawners to provide pending_messages and inform users.

### PR DESCRIPTION
**Pull request by clkao**
This is prereq for https://github.com/jupyterhub/kubespawner/issues/148 and https://github.com/jupyterhub/kubespawner/issues/141

### CODE REVIEW

File: share/jupyterhub/templates/spawn_pending.html [hunk 1]
```diff
@@ -12,7 +12,12 @@
       <p><i class="fa fa-spinner fa-pulse fa-fw fa-3x" aria-hidden="true"></i></p>
       <a role="button" id="refresh" class="btn btn-lg btn-primary" href="#">refresh</a>
     </div>
-  </div>
+    <pre>
+      {% for m in messages %}
```
Comment (minrk) [inline share/jupyterhub/templates/spawn_pending.html]: This probably shouldn't be in `pre`. They should either be in `<p>` tags or passed as-is for HTML messages.

**Comment by minrk** [comment]
Thanks for the PR! This doesn't provide the basis needed for either of the above linked issues, though. https://github.com/jupyterhub/kubespawner/issues/141 is for failed launches, which don't show on the pending-spawn page. And jupyterhub/kubespawner#148 asks for a mechanism for streaming events, not a static spawner-provided message that doesn't change. While this would improve the pending-message case somewhat because a Spawner could provide a static description describing what's going on, if we want to have a full event-stream of messages, then I think we probably don't want to implement two versions of the same thing for Spawners to deal with.

**Comment by clkao** [comment]
For failed launches, I was mostly referring to slow span that eventually timed out. Otherwise Exceptions are already displayed. So the launch progress message was sufficient for my use case.

**Comment by clkao** [comment]
For streaming messages we probably need to introduce websocket to the codebase first.  I thought we can start with static messages, and maybe a watcher for the message array that emits events over websocket in the future.

**Comment by minrk** [comment]
The reason I'm hesitant about adding this message API is that it doesn't address what's requested in jupyterhub/kubespawner#148 and if we do implement a streaming message API, that would immediately replace this API. I'm hesitant to introduce a new API when we already have a plan to deprecate it immediately. Since the structure (necessarily async, probably an async generator) would be quite different, I'm not sure anything in this API could be reused.

**Comment by minrk** [comment]
#1771 implements the async generator and event-stream

**Comment by minrk** [comment]
closing in favor of #1771

**Pull request closed without merging**